Do not generate default methods for java.lang.Object methods

In Java8 it's not allowed to define a default method for methods found in java.lang.Object.
If you try to do it, you will run into an error message like below:
```
[ERROR] /home/reegnz/github/virtual-field-example/target/generated-sources/annotations/java/util/VirtualList.java:[93,17] default method hashCode in interface java.util.VirtualList overrides a member of java.lang.Object
```
Reasoning by Brian Goetz also gives some ideas why allowing default method implementations for these methods might be problematic:
http://mail.openjdk.java.net/pipermail/lambda-dev/2013-March/008435.html

The problem is that you are free to redefine the same method signature in interfaces as regular abstract methods. So blindly copying all methods from the parent interface and defining a default method for all of them won't work.

Task:
Make sure to only define default methods for interface methods not redefining the method signature of java.lang.Object methods.

--- a/src/main/java/reegnz/processor/VirtualFieldProcessor.java
+++ b/src/main/java/reegnz/processor/VirtualFieldProcessor.java
@@ -51,6 +51,7 @@ public class VirtualFieldProcessor extends AbstractProcessor {
 	private Elements elementUtils;
 	private Filer filer;
 	private Messager messager;
+	private Set<String> objectMethods;
 
 	@Override
 	public SourceVersion getSupportedSourceVersion() {
@@ -70,6 +71,23 @@ public class VirtualFieldProcessor extends AbstractProcessor {
 		elementUtils = processingEnv.getElementUtils();
 		filer = processingEnv.getFiler();
 		messager = processingEnv.getMessager();
+		objectMethods = getObjectMethodSignatures();
+	}
+
+	/**
+	 * Gets all java.lang.Object methods that are are possible to override.
+	 * @param name
+	 * @return
+	 */
+	private Set<String> getObjectMethodSignatures() {
+		return elementUtils.getTypeElement("java.lang.Object")
+				.getEnclosedElements().stream()
+				.filter(e -> e.getKind().equals(ElementKind.METHOD))
+				.map(ExecutableElement.class::cast)
+				.filter(e -> !e.getModifiers().contains(Modifier.STATIC))
+				.filter(e -> !e.getModifiers().contains(Modifier.FINAL))
+				.map(ExecutableElement::toString)
+				.collect(Collectors.toSet());
 	}
 
 	@Override
@@ -98,6 +116,7 @@ public class VirtualFieldProcessor extends AbstractProcessor {
 		List<ExecutableElement> methods = element.getEnclosedElements().stream()
 				.filter(this::isMethod)
 				.map(ExecutableElement.class::cast)
+				.filter(e -> !objectMethods.contains(e.toString()))
 				.filter(this::isNotStatic)
 				.collect(toList());
 		generate(element, methods);

